### PR TITLE
coverage: Clean up `mcdc_bitmap_bytes` and `conditions_num`

### PR DESCRIPTION
**Pull request by Zalathar**
This is a combination of two mostly-separate MC/DC coverage cleanups that would conflict with each other, plus some extra tests that appeared along the way.

The first change is to stop recomputing `mcdc_bitmap_bytes` in the query that produces `CoverageIdsInfo`. This appears to have been inspired by how we were already computing `max_counter_id`, but there's an important difference between the two cases.

When computing `max_counter_id`, the highest counter ID seen might be less than the highest ID used during MIR instrumentation, because some counter-increment statements might have been removed by MIR optimizations.

But for the recomputation used for `mcdc_bitmap_bytes`, that's impossible, because both computations are based on pre-optimization info. So there's no need to recompute the exact same value, when it can't have changed

---

The second change is to rename `conditions_num` to `num_conditions`, to make it clearer that this refers to a *number of conditions*, not some kind of ID number.

Because this change touched the compiler warning for a decision containing too many conditions, I also noticed that we didn't have any tests for that warning.

(It now seems a bit strange to me that this is a compiler warning, not a lint, because it can't be silenced or denied by the usual mechanisms for controlling lints. But I consider that change to be beyond the scope of this PR.)

### CODE REVIEW

**Comment by Zalathar** [comment]
It might be possible to adjust the calculation of MC/DC bitmap bytes so that it takes into account which MC/DC coverage statements have survived MIR optimizations, to potentially use fewer bytes.

But in the current state of the MC/DC implementation, I don't think the extra complexity is worth it.

(The main reason we do this sort of thing for the existing coverage statements is that things already worked that way before I started modifying it, and I didn't want to regress that particular property of the instrumented output.)

**Comment by Zalathar** [comment]
Looks like the overflow bug already existed before this PR; it just now gets exercised by the newly-added test (but only in compiler builds with overflow checks).

The overflow itself is pretty benign in known cases, since we overflow by -1 and then immediately add 1 again to get back to 0. But it also prevents us from detecting more serious overflow bugs.

File: compiler/rustc_mir_build/src/build/coverageinfo/mcdc.rs [hunk 1]
```diff
@@ -207,27 +207,27 @@ impl MCDCInfoBuilder {
...
                     unreachable!("Decision with no condition is not expected");
                 }
-                1..=MAX_CONDITIONS_NUM_IN_DECISION => {
+                1..=MAX_CONDITIONS_IN_DECISION => {
                     self.decision_spans.push(decision);
                 }
                 _ => {
                     // Do not generate mcdc mappings and statements for decisions with too many conditions.
-                    let rebase_idx = self.branch_spans.len() - decision.conditions_num + 1;
+                    let rebase_idx = self.branch_spans.len() - decision.num_conditions + 1;
```
Comment (RenjiSann) [inline compiler/rustc_mir_build/src/build/coverageinfo/mcdc.rs:219]:
```suggestion
                    let rebase_idx = self.branch_spans.len() - (decision.num_conditions + 1);
```

Wouldn't adding parenthesis be the easiest fix ?

**Comment by RenjiSann** [comment]
> Looks like the overflow bug already existed before this PR; it just now gets exercised by the newly-added test (but only in compiler builds with overflow checks).
> 
> The overflow itself is pretty benign in known cases, since we overflow by -1 and then immediately add 1 again to get back to 0. But it also prevents us from detecting more serious overflow bugs.

Maybe rewrite the operation `+ 1 - num_conditions` to avoid the underflow. It looks weird, but it's probably the easiest and most readable fix, I guess ?

**Comment by Zalathar** [comment]
Closing this in favour of #125754.

**Pull request closed without merging**